### DISCUSSION TITLE
[General] ETH <-> zkSync crosschain messaging errors

### DISCUSSION THREAD

**Discussion by kss-fluency** [zkSync-Community-Hub/zksync-developers#116] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

hi devs! I am using zkSync's x-chain protocol (as per https://era.zksync.io/docs/dev/tutorials/cross-chain-tutorial.html ) to send a message from a Ethereum smart contract to a zkSync smart contract. the idea is to have a NFT smart contracts on Ethereum and a Mirror smart contract in zkSync that will reflect the Ethereum contract state (specifcally token owners)

### What is the specific issue or error you're encountering?

No specific issue. But I have some questions:

- what happens if sending the message fails - for example by insufficient ether sent in msg.value to cover the zkSync update?
- the IZkSync.requestL2Transaction call produces a txHash. how I can utilise it? can it be used to retry a failed message?
- is there a way to estimate how much msg.value do I need to send alongside IZkSync.requestL2Transaction onchain, while executing a Etherum smart contract ?

### Can you share the error messages or logs you're receiving, if any?

N/A

### Have you made any recent changes to the contract before encountering this issue?

N/A

### Are there any external libraries or contracts that your contract interacts with?

N/A

### Can you provide the relevant portions of your contract code where the issue is occurring?

N/A

### Have you tried to isolate the problem, and if so, what were the results?

N/A

### What steps have you already taken to try to resolve the issue?

I tried to find responses to my questions in the zkSync documentation, but they don't seem to be there

### Additional Details

I am building CommonWealth (https://www.common-wealth.io/)

**Comment by vladbochok** [comment, marked as answer]
Hello @kss-fluency, 

> what happens if sending the message fails - for example by insufficient ether sent in msg.value to cover the zkSync update?

zkSync smart contract require some fees in advance to cover the cost of processing transaction. If there is not enough value, L1 contract will revert execution and nothing will happen.  If L1 -> L2 transaction successfully submitted and then failed on L2 by any reason (revert in the execution as not enough `msg.value`) then user will receive all funds on L2 account. 

> the IZkSync.requestL2Transaction call produces a txHash. how I can utilise it? can it be used to retry a failed message?

You can't retry a failed L1 -> L2 transaction, the only way to retry it is to send another L1 -> L2 transaction. User woudn't lose any funds except the fees.

> is there a way to estimate how much msg.value do I need to send alongside IZkSync.requestL2Transaction onchain, while executing a Etherum smart contract ?

Yes, you can find `l2TransactionBaseCost` function.

```
function l2TransactionBaseCost(
        uint256 _gasPrice,
        uint256 _l2GasLimit,
        uint256 _l2GasPerPubdataByteLimit
    )
```

- https://github.com/matter-labs/era-contracts/blob/main/ethereum/contracts/zksync/facets/Mailbox.sol#L150

**Comment by amyliasnow** [comment]
Sure, here are answers to your questions:

If sending the message fails due to insufficient ether sent in msg.value to cover the zkSync update, the transaction will revert, and any changes made before the failure will be reverted as well. It's essential to ensure that enough ether is sent along with the transaction to cover the fees.

The txHash returned by IZkSync.requestL2Transaction is the hash of the zkSync transaction. You can use it to track the status of the transaction on the zkSync network. If the message fails, you can't directly retry it using the same txHash. You would need to create a new transaction with updated parameters and send it again.

To estimate how much msg.value you need to send alongside IZkSync.requestL2Transaction, you can use the estimated gas cost of the transaction and multiply it by the current gas price. Additionally, you need to consider any additional fees required by zkSync, such as the transfer fee. You can adjust the msg.value accordingly based on these factors.

Hope this helps! Let me know if you have any further questions.

**Comment by ryangrullon** [comment]
Certainly! Here are brief answers to your questions:

If sending the message fails due to insufficient ether sent in msg.value to cover the zkSync update, the transaction will revert and any changes made up to that point will be reverted as well.

The txHash produced by IZkSync.requestL2Transaction can be utilized to monitor the status of the transaction on zkSync's Layer 2. You can use this hash to check whether the transaction has been successfully executed or if it has failed.

Estimating the msg.value required to cover the zkSync update can be challenging as it depends on various factors such as gas prices and the complexity of the transaction. One approach is to simulate the transaction using a local development environment or testnet to get an estimate of the gas cost, and then adjust the msg.value accordingly. Additionally, you can implement mechanisms in your contract to dynamically adjust the msg.value based on real-time gas prices.

**Comment by reginaldolivas** [comment]
For questions regarding handling failed messages, utilizing txHash, and estimating msg.value for zkSync transactions, 
it's recommended to directly reach out to the zkSync support team or community forums for precise guidance and assistance.

**Comment by postm4ltolin** [comment]
If sending the message fails due to insufficient ether sent in msg.value to cover the zkSync update, the transaction will fail and the message will not be sent. The txHash produced by the IZkSync.requestL2Transaction call can be used to track the status of the transaction and determine if it was successful or not. It can also be used to retry a failed message. To estimate how much msg.value to send alongside IZkSync.requestL2Transaction, you can use the gas limit and gas price of the transaction as well as the current exchange rate between ETH and zkETH. Additionally, you can test with small amounts of ether first to determine the minimum required amount for successful execution.

**Comment by cioprogephen** [comment]
The user is using zkSync's x-chain protocol to send a message from an Ethereum smart contract to a zkSync smart contract for the purpose of reflecting the Ethereum contract state. The user has some questions regarding the process, such as what happens if sending the message fails, how to utilize the txHash produced by the IZkSync.requestL2Transaction call, and how to estimate the amount of msg.value to send alongside the call. The user has not encountered any specific issues or errors, and no relevant portions of their contract code are provided. The user has tried to find answers in the zkSync documentation but has not found them. The user is building CommonWealth.
